aips/aip-56.md

```diff
@@ -0,0 +1,213 @@
+```
+
+
+#### Runtime Evaluation
+
+At runtime, a stack of saved access specifiers is maintained as well as an active current access specifier. When a function which has access specifiers is entered, the current active set is saved to the stack. Then its value is _joined_ (the `*` operator above) with that of the current function. If the result of the join does not subsume the specifier of the called function, execution aborts:
+
+```
+access_stack.push(active_accces)
+active_access := active_access * function_access
+let call_allowed := active_access.subsumes(function_access)
+if !(active_access >= function_access) {
+    abort
+}
+```
+Notice that we cannot just always abort when entering a function which potentially has disallowed accesses for two reasons:
+
+- The function many not have any specifiers at all. In this case, the `active_access` stays unmodified at function entry, and the actual accesses are checked as they occur.
+- The implementation of subsumption (`>=`) is allowed to be an over-approximation. This stems from the presence of negation in specifiers, for which subsumption check is hard. Thus, we may not be able to decide `active_access >= function_access`. This case is semantically still sound because, similar with functions with no specifiers at all, accesses are still checked as they occur.
+
+During execution, whenever a resource is accessed (move_to, move_from, exists, borrow_global, borrow_global_mut), this operation is checked against the currently active access specifier, and execution aborts if the access is not allowed.
+
+Note that unless trust boundaries are concerned (transactions and public functions), we rarely expect users to write access specifiers, so the stack is not expected to become very deep in average. 
+
+#### Gas Cost
+
+Access specifiers need to be metered, also to prevent their misuse. This is different from how Aptos currently treats paranoid mode. Because this feature adds additional utility to users, metering is justified.
+
+#### Compatibility
+
+Since access specifiers are encoded in the `FunctionHandle` instances which are copied from imported modules, they get outdated when the dependency is upgraded. The compatibility rule for upgrade is that access specifiers can only be constrained, that is `S_v1 >= S_v2 >= ...`
+
+
+## Reference Implementation
+
+See [PR #10480](https://github.com/aptos-labs/aptos-core/pull/10480) for the compiler implementation.
+
+See [PR #10544](https://github.com/aptos-labs/aptos-core/pull/10544) for the VM implementation.
+
+## Testing
+
+The behavior of access specifiers can be well tested with unit tests. Significant coverage is expected. 
+
+
+## Risks and Drawbacks
+
+Risks specifically include developer adoption and engineering complexity. This feature need to be well documented with intuitive examples. We may also build tools which allow to derive access specifiers for public functions automatically, easing the effort.
```

Comment:
If it turns out that multiple functions are being annotated with the same (long list of) access specifiers, we could consider adding an "access specifier alias" (i.e., an explicitly declared name for a list of access specifiers, that could be used in lieu of the full list).

Reply:
Yes. Though I would expect only public functions are annotated, the rest inferred.